server/src/modules/jam.tsx

```diff
@@ -0,0 +1,234 @@
+        console.log('ws connected');
+        ws.send(JamView().toString());
+
+        ws.on('message', (msg) => {
+            console.log(msg);
+        });
+
+        ws.on('close', () => {
+            console.log('ws closed');
+            connectedSockets.splice(connectedSockets.indexOf(ws), 1);
+        });
+    });
+};
+
+const refreshAll = () => {
+    for (const ws of connectedSockets) {
+        ws.send(JamView().toString());
+    }
+};
+
+jamRouter.post<undefined, JSX.Element, undefined, {chord: string}>(
+    ROUTES.JAM_ACTIONS_ADD_CHORD,
+    (req, res) => {
+        const {chord} = req.query;
+        jamState.selectedChords.push(chord);
+
+        res.send(JamView());
+        refreshAll();
+    },
+);
+
+jamRouter.post<undefined, JSX.Element>(
+    ROUTES.JAM_ACTIONS_NEW_JAM,
+    (req, res) => {
+        jamState.selectedChords = [];
+
+        res.send(JamView());
+        refreshAll();
+    },
+);
+
+jamRouter.post<undefined, JSX.Element, {key: string}>(
+    ROUTES.JAM_ACTIONS_CHANGE_KEY,
+    (req, res) => {
+        const {key} = req.body;
+        jamState.key = key;
+
+        res.send(JamView());
+        refreshAll();
+    },
+);
+
+export const renderJamPage = () => {
+    return JamPage();
+};
+
+export const JamPage = () => {
+    return (
+        <>
+            <div hx-ws='connect:/jam/ws' />
+            <JamView />
+        </>
+    );
+};
+
+export const JamView = () => {
+    const selectedChords = jamState.selectedChords;
+    const key = jamState.key;
+    const availableChords = getChordsFromKeyRoot(key);
+
+    return (
+        <div
+            id='jam-view'
+            hx-ws='message:replaceOuterHTML'
+            style={{textAlign: 'center'}}
+        >
+            <KeyName root={key} quality='maj' />
+            <ChooseKeySelectBox key={key} />
+            <NewJamButton />
+            <ChordSelectorSection availableChords={availableChords} />
+            <DraftViewSection selectedChords={selectedChords} />
+        </div>
+    );
+};
+
+const DraftViewSection = (props: {selectedChords: string[]}) => {
+    return (
+        <div class='drafted-chords'>
+            {props.selectedChords.map((chordName) => (
+                <span class='displayed-chord' style={{margin: '20px'}}>
+                    {chordName}
+                </span>
+            ))}
+        </div>
+    );
+};
+
+const ChordSelectorSection = (props: {availableChords: string[]}) => {
+    console.log(props.availableChords);
+    return (
+        <div class='chord-buttons'>
+            {props.availableChords.map((chordName) => {
+                const params = new URLSearchParams();
+                params.append('chord', chordName);
+                return (
+                    <button
+                        type='button'
+                        class='chord-button'
+                        hx-post={`${
+                            ROUTES.JAM_ACTIONS_ADD_CHORD
+                        }?${params.toString()}`}
```

Comment:
This is disgusting, and exactly why I suggest we use functions passed into views to avoid this exact sort of thing, in the comment below